### ISSUE TITLE
Feature Proposal: Allow specifying fallback component for error boundary.

### ISSUE THREAD

**Issue by foxyblocks** [bugsnag/bugsnag-react#3]
Currently if there is an error during rendering and it get's caught by the `bugsnag-react` error boundary it will render a blank screen. According to [the react docs](https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html#introducing-error-boundaries) the best practice for error boundaries is the render a fallback component when an error is present. In order to do this with `bugsnag-react` one would have to stop using the built in error boundary or copy and paste the source to modify it. 

We should make the generated error boundary component accept a prop for fallback component that it would render in the event of an error. 

The usage docs would need to be updated to something like the following:
```javascript
import bugsnag from 'bugsnag-js'
const bugsnagClient = bugsnag('API_KEY')

import ReactDOM from 'react-dom'
import React from 'react'
import createPlugin from 'bugsnag-react'

function ErrorScreen() {
  return (
    <div>Something went wrong, we have been notified of the problem. Try refreshing your browser window.</div>
  )
}

// wrap your entire app tree in the ErrorBoundary provided
const ErrorBoundary = bugsnagClient.use(createPlugin(React))
ReactDOM.render(
  <ErrorBoundary fallbackComponent={ErrorScreen}>
    <YourApp />
  </ErrorBoundary>,
  document.getElementById('app')
)
```

**Comment by bengourley** [comment]
I like this proposal and I think we should do it ✨

I just wanted to respond to this comment you made because I don't think it's true

> In order to do this with bugsnag-react one would have to stop using the built in error boundary or copy and paste the source to modify it.

Here's a quick sketch of how I envisaged doing this by extending the Bugsnag ErrorBoundary class…

```js
const ErrorBoundary = bugsnagClient.use(createPlugin(React))
class MyErrorBoundary extends ErrorBoundary {
  componentDidCatch(err, info) {
    super.componentDidCatch(err, info)
    this.setState({ error: true })
  }
  render () {
    if (this.state.error) return <div>Error etc.</div>
    return super.render()
  }
}
```

However this requires more thought and effort than your proposal. My only concern is that to implement it we'd need to call `setState()` on the component. Is this an ok thing to do? Might it interfere with data structures in people's applications?

**Comment by foxyblocks** [comment]
Typically the problem with data structure interference is one reason (among many) why react developers don't like the idea of extending third party classes and instead prefer function composition. It could be that rather than creating an error boundary, the bugsnag react library could just be an enhancement that adds component stacktraces. Then it would be used like so:

```js
const bugsnagReact = bugsnagClient.use(createPlugin())
class MyErrorBoundary extends React.PureComponent {
  componentDidCatch(err, info) {
    bugsnagReact.componentDidCatch(err, info);
    this.setState({ error: true })
  }
  render () {
    if (this.state.error) return <div>Error etc.</div>
    return this.props.children;
  }
}
```

This would provide all the same benefits while providing the most flexibility.

**Comment by snmaynard** [comment]
I think adding a beforeNotify and a fallbackComponent prop to the boundary would be good. That way you can do complex changes or mutate the report in the beforeNotify but also just render a different component easily if thats all you need. I think this is better than the approach above while still being composition. Thoughts?

```javascript
errorThrown(err, report) {
  this.setState({error:true})
  report.addMetadata({state: this.state})
  return true
}
render() {
  if(this.state.error) return <div>Error etc.</div>
  return (<BugsnagErrorBoundary beforeNotify={this.errorThrown}>
    //Other stuff
  </BugsnagErrorBoundary>)
}
```

**Comment by snmaynard** [comment]
https://github.com/bvaughn/react-error-boundary here is someone doing something similar

**Comment by bengourley** [comment]
Implemented by @martin308 and published in v1.1.1 🎉
